### PR TITLE
refactor: restructured `impl`, changed docs testing

### PR DESCRIPTION
**Pull request by g3ner1c**
# Changes

## Refactored `impl` [as per discord proposal](https://discord.com/channels/929221894316625960/929221894316625963/982371550302969906)

- Old:
```
 impl
├── custom.py
└── modern.py
```
- New:
```
 impl
├── scorer.py
├── custom.py
├── gravity.py
├── presets.py
├── queue.py
└── rotation.py
```
- Individual game mechanics are now in their own seperate modules, while presets are in `presets.py`:
	- `impl.modern.ModernEngine` -> `impl.presets.ModernEngine`
	- `impl.modern.InfinityGravity` -> `impl.gravity.InfinityGravity`
	- `impl.modern.SevenBag` -> `impl.queue.SevenBag`
	- `impl.modern.SRS` -> `impl.rotation.SRS`
	- `impl.modern.Scorer` -> `impl.scorer.GuidelineScorer`

## `tox.ini`

- Changed `whitelist_externals` to `allowlist_externals` as [the former is deprecated](https://tox.wiki/en/latest/config.html#:~:text=whitelist_externals%20has%20the%20same%20meaning%20and%20usage%20as%20allowlist_externals%20but%20it%20is%20now%20deprecated.)

- `testenv:docs` now deletes the entire `docs/api` directory before regenerating apidocs in case new files get added or deleted in source code

## `.github/workflows/test.yml`

- `git diff-index` with `--exit-code` instead of `--quiet` achieves the same function without supressing output
- Ok this ones a bit weird but for some reason after recreating the apidocs with the exact same contents `git diff-index` says there are modified files when the rest of git doesn't agree:
```
$ tox -e docs
docs inst-nodeps: /home/sky/python/g3ner1c-python-tetris/.tox/.tmp/package/1/tetris-0.6.0.tar.gz
...
$ git diff-index --exit-code HEAD
:100644 100644 9dd1456360c6e40d7dce8fb7eb53babf2470fefa 0000000000000000000000000000000000000000 M      docs/api/tetris.engine.rst
:100644 100644 fd13b0d74e090544c733a6adf79b7e9746478182 0000000000000000000000000000000000000000 M      docs/api/tetris.game.rst
:100644 100644 391cd06a5b7d962ccad3d8dbd394f72a426fde4c 0000000000000000000000000000000000000000 M      docs/api/tetris.impl.custom.rst
:100644 100644 214371d5bf61562207a46e3b2b59213fe9dc10b3 0000000000000000000000000000000000000000 M      docs/api/tetris.impl.gravity.rst
:100644 100644 b830ab707b25e5e2af35637b8c30f4bdbd9f0529 0000000000000000000000000000000000000000 M      docs/api/tetris.impl.presets.rst
:100644 100644 f18bac554bf5a516efbf5be60695e7f9f719799f 0000000000000000000000000000000000000000 M      docs/api/tetris.impl.queue.rst
:100644 100644 286cc53df281627ff5878ba896b60c9107ec7d65 0000000000000000000000000000000000000000 M      docs/api/tetris.impl.rotation.rst
:100644 100644 818bf07d8f3d53c815e31512abe8390d7e8c62a3 0000000000000000000000000000000000000000 M      docs/api/tetris.impl.rst
:100644 100644 40e2d93d9e57663f1b4123a66e06b2de0f25a8ca 0000000000000000000000000000000000000000 M      docs/api/tetris.impl.scorer.rst
:100644 100644 849f3d0aaae4d42fd490513ba2a50e5d456ef35a 0000000000000000000000000000000000000000 M      docs/api/tetris.rst
:100644 100644 8131e1ec01e86d6a1b65213b21d5de2b56ab4d6d 0000000000000000000000000000000000000000 M      docs/api/tetris.types.rst
$ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
``` 
However, running `git status` before `git diff-index` resolves this for some reason idk why lol ¯\\\_(ツ)_/¯
```
$ tox -e docs
docs inst-nodeps: /home/sky/python/g3ner1c-python-tetris/.tox/.tmp/package/1/tetris-0.6.0.tar.gz
...
$ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
$ git diff-index --exit-code HEAD
*prints nothing, exit code 0*
```

### CODE REVIEW

**Comment by dzshn** [comment]
does `git diff --exit-code` instead of `diff-index` have this quirk? I think the latter is a lower level command

File: tetris/impl/custom.py [hunk 1]
```diff
@@ -53,10 +56,10 @@ class CustomEngine(Engine):
...
     def __init__(
         self,
-        gravity: Gravity = modern.InfinityGravity,
-        queue: Queue = modern.SevenBag,
-        rotation_system: RotationSystem = modern.SRS,
-        scorer: Scorer = modern.Scorer,
+        gravity: Gravity = gravity.InfinityGravity,
+        queue: Queue = queue.SevenBag,
+        rotation_system: RotationSystem = rotation.SRS,
+        scorer: Scorer = scorer.GuidelineScorer,
```
Comment (dzshn) [inline tetris/impl/custom.py:62]:
```suggestion
        gravity: Gravity = InfinityGravity,
        queue: Queue = SevenBag,
        rotation_system: RotationSystem = SRS,
        scorer: Scorer = GuidelineScorer,
```

File: tetris/impl/custom.py [hunk 2]
```diff
@@ -10,7 +10,10 @@
 from tetris.engine import Queue
 from tetris.engine import RotationSystem
 from tetris.engine import Scorer
-from tetris.impl import modern
+from tetris.impl import gravity
+from tetris.impl import queue
+from tetris.impl import rotation
+from tetris.impl import scorer
```
Comment (dzshn) [inline tetris/impl/custom.py:16]:
```suggestion
from tetris.impl.gravity import InfinityGravity
from tetris.impl.queue import SevenBag
from tetris.impl.rotation import SRS
from tetris.impl.scorer import GuidelineScorer
```
we only use single objects from these, I think this is less messy

**Comment by g3ner1c** [comment]
seems not!

File: tetris/impl/rotation.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        piece.minos = self.shapes[piece.type][piece.r]
+
+
+_Tetrio_override = {
+    (0, 2): ((-1, +0), (-1, +1), (-1, -1), (+0, +1), (+0, -1)),
+    (1, 3): ((+0, +1), (-2, +1), (-1, +1), (-2, +0), (-1, +0)),
+    (2, 0): ((+1, +0), (+1, -1), (+1, +1), (+0, -1), (+0, +1)),
+    (3, 1): ((+0, -1), (-2, -1), (-1, -1), (-2, +0), (-1, +0)),
+}
```
Comment (dzshn) [inline tetris/impl/rotation.py:139]: I just realised this is extraneous from [when there was a `TetrioSRS` object](https://github.com/dzshn/python-tetris/blob/b995c9f0dcef9cd4a9823358665112b31ced4d4a/tetris/engine/rotation.py#L140-L144), I suggest to either remove this or re-add that

File: tetris/impl/presets.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+from tetris import Engine
+
+if TYPE_CHECKING:
+    from tetris import BaseGame
+
+from tetris.impl import gravity
+from tetris.impl import queue
+from tetris.impl import rotation
+from tetris.impl import scorer
```
Comment (dzshn) [inline tetris/impl/presets.py:15]:
```suggestion
from tetris.impl.gravity import InfinityGravity
from tetris.impl.queue import SevenBag
from tetris.impl.rotation import SRS
from tetris.impl.scorer import GuidelineScorer
```
the same applies here

File: tetris/impl/presets.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        return queue.SevenBag(seed=game.seed)
+
+    def scorer(self, game: BaseGame) -> scorer.GuidelineScorer:  # noqa: D102
+        return scorer.GuidelineScorer()
+
+    def rotation_system(self, game: BaseGame) -> rotation.SRS:  # noqa: D102
+        return rotation.SRS(board=game.board)
+
+    def gravity(self, game: BaseGame) -> gravity.InfinityGravity:  # noqa: D102
+        return gravity.InfinityGravity(game=game)
```
Comment (dzshn) [inline tetris/impl/presets.py:31]:
```suggestion
    def queue(self, game: BaseGame) -> SevenBag:  # noqa: D102
        return queue.SevenBag(seed=game.seed)

    def scorer(self, game: BaseGame) -> GuidelineScorer:  # noqa: D102
        return scorer.GuidelineScorer()

    def rotation_system(self, game: BaseGame) -> SRS:  # noqa: D102
        return rotation.SRS(board=game.board)

    def gravity(self, game: BaseGame) -> InfinityGravity:  # noqa: D102
        return gravity.InfinityGravity(game=game)

```

**Review by dzshn — APPROVED** [review]
thank you for this PR! I left a couple of suggestions here before merging

File: tetris/impl/rotation.py (hunk 3 above)
Comment (g3ner1c) [reply tetris/impl/rotation.py:139]: we can probably keep this in for now as I do want to add that in later along with [all the other TETR.IO kick tables](https://user-images.githubusercontent.com/74469257/172017621-313d15ce-edcd-4435-8df5-215dc1da3963.png)

File: tetris/impl/presets.py (hunk 4 above)
Comment (g3ner1c) [reply tetris/impl/presets.py:15]: for this one I intentionally kept the scope larger so later on when more presets get added you don't have to individually specify each object

File: tetris/impl/presets.py (hunk 5 above)
Comment (g3ner1c) [reply tetris/impl/presets.py:31]: same as the one above

**Review by dzshn — APPROVED** [review]

**Pull request merged**